assert2-macros/src/let_assert.rs

```diff
@@ -0,0 +1,134 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::punctuated::Punctuated;
+
+use crate::spanned_to_string;
+use crate::FormatArgs;
+
+pub struct Args {
+	macro_name: syn::Expr,
+	pattern: syn::Pat,
+	eq_token: syn::token::Eq,
+	expression: syn::Expr,
+	format_args: Option<FormatArgs>,
+}
+
+pub fn let_assert_impl(args: Args) -> TokenStream {
+	let Args {
+		macro_name,
+		pattern,
+		eq_token,
+		expression,
+		format_args,
+	} = args;
+
+	let placeholders = collect_placeholders(&pattern);
+
+	let mut use_placeholders = TokenStream::new();
+	for placeholder in &placeholders {
+		use_placeholders.extend(quote! {
+			&#placeholder,
+		});
+	}
+	let use_placeholders = quote! {
+		let _ = (#use_placeholders);
+	};
+
+	let placeholders = puncuate_idents(placeholders);
+
+	let pat_str = spanned_to_string(&pattern);
+	let expr_str = spanned_to_string(&expression);
+	let custom_msg = match format_args {
+		Some(x) => quote!(Some(format_args!(#x))),
+		None => quote!(None),
+	};
+
+	quote! {
+		let (#placeholders) = {
+			let value = #expression;
+			if let #pattern #eq_token &value {
+				#use_placeholders
+				if let #pattern #eq_token value {
+					(#placeholders)
+				} else {
+					panic!("{}: second pattern match failed, please report this at https://github.com/de-vri-es/assert2-rs/issues/new/");
+				}
+			} else {
```

Comment:
Hmm, I figured otherwise the value would always be consumed, making it impossible to borrow it in the error case. I never considered the possibility that the value is still alive in the `else`. If so, that is pretty cool. Will test soon.